planning/louketo-proxy-plans.md

```diff
@@ -0,0 +1,420 @@
+
+* OpenID Connect Front-Channel Logout
+
+* OpenID Connect Back-Channel Logout
+
+* OpenID Connect Dynamic Registration
+
+* JSON Web Signature (JWS)
+
+* JSON Web Encryption (JWE)
+
+* JSON Web Key (JWK)
+
+* OAuth 2.0 Token Introspection
+
+* OAuth 2.0 Security Best Current Practice
+
+* Financial-grade API - Part 1: Read-Only API Security Profile
+
+* Financial-grade API - Part 2: Read and Write API Security Profile
+
+* User Managed Access 2.0
+
+* OAuth 2.0 Mutual TLS Client Authentication and Certificate Bound Access Tokens
+
+# Review options
+
+## Envoy
+
+[https://www.envoyproxy.io/](https://www.envoyproxy.io/)
+
+[https://github.com/envoyproxy/envoy](https://github.com/envoyproxy/envoy)
+
+### Overview
+
+Envoy is a general purpose proxy. At least 90% of what it provides we don’t need, and it only provides 10% of what we need. JWT processing is very limited and there is currently no support for OpenID Connect or OAuth 2.0. As it is written in C++ there would be a learning curve for the team to be able to contribute to the project. It would also be fairly complex to productize and deliver a supported distribution to customer. With so many additional features that are not needed, support may also be problematic.
+
+### Project
+
+* Large community (11k starts, 436 contributors)
+
+* ASL2 license
+
+* Implemented in C++
+
+### Notes
+
+* 90% of what it does we don’t need, and it only does 10% of what we need
+
+* Delivering a supported distribution may be complicated due to the large amount of unneeded features
+
+* Configuration is complex due to the amount of features
+
+* Getting a basic example up and running is not trivial
+
+* No OpenID Connect or OAuth support today
+
+    * Open issue for OAuth HTTP filter [https://github.com/envoyproxy/envoy/issues/8571](https://github.com/envoyproxy/envoy/issues/8571)
+
+    * Indication that there is interest in adding
+
+* Basic authentication with JWT Authentication HTTP filter
+
+    * Basic JWT checks
+
+    * Fetch keys from JWKS
+
+        * Requires setting up a cluster
+
+        * Looks like keys are cached for a fixed interval and there is no support for updating when keys are rotated other than waiting for cache to expire
+
+    * Ability to forward JWT upstream
+
+    * Support passing JWT as query parameter, which is not recommended unless JWT is limited to a specific request such as downloading a file
+
+    * Limited claim processing
+
+* Written in C which may impact ability to contribute missing capabilities
+
+* Can be extended with C
+
+    * Not sure if it’s possible to add custom claim processing without replacing the JWT filter
+
+    * Mentions of WebAssembly support in the future [https://github.com/envoyproxy/envoy/issues/4272](https://github.com/envoyproxy/envoy/issues/4272)
+
+    * If we need custom extensions it may be complicated to maintain and deliver a separate distribution
+
+## Pusher OAuth2 Proxy
```

Comment:
@stianst Also I think make sense to remove this section, you think otherwise.

Reply:
+1